cprover_bindings/src/goto_program/symbol.rs

```diff
@@ -46,6 +48,14 @@ pub struct Symbol {
     pub is_thread_local: bool,
     pub is_volatile: bool,
     pub is_weak: bool,
+
+    /// This flag marks a variable as constant (IrepId: `ID_C_constant`).
+    ///
+    /// In CBMC, this is a property of the type or expression. However, we keep it here to avoid
+    /// having to propagate the attribute to all variants of `Type` and `Expr`.
+    ///
+    /// During contract verification, CBMC will not havoc static variables marked as constant.
```

Comment:
Should we distinguish promoted constants from user-declared symbols declared as constants ? (the first type are really literal constants, the second one may be modified through unsafe) ?

Reply:
No, I don't think so. We should however handle `UnsafeCell`. We could for now check if the type has one. If so, we keep it as mutable.

Note that mutating a constant is immediate UB.

>Mutating immutable bytes. All bytes inside a [const](https://doc.rust-lang.org/reference/items/constant-items.html) item are immutable. The bytes owned by an immutable binding are immutable, unless those bytes are part of an [UnsafeCell<U>](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html).

https://doc.rust-lang.org/reference/behavior-considered-undefined.html

Reply:
Note that the following program crashes:
```rust
static STATIC_CONST: u32 = 0;

fn check_static() {
    assert_eq!(STATIC_CONST, 0);
}

fn assign(ptr: *mut u32) {
    unsafe { *ptr = 10 };
}

fn main() {
    let ptr = &STATIC_CONST as *const _ as *mut _;
    assign(ptr); // -- Doing the assignment here triggers a compilation error.
    check_static();
}
```